vendor/github.com/prometheus/tsdb/wal/wal.go

```diff
@@ -899,6 +899,210 @@ func (r *Reader) Offset() int64 {
 	return r.total
 }
 
+// NewLiveReader returns a new live reader.
+func NewLiveReader(r io.Reader) *LiveReader {
+	return &LiveReader{rdr: r}
+}
+
+// Reader reads WAL records from an io.Reader. It buffers partial record data for
+// the next read.
+type LiveReader struct {
+	rdr        io.Reader
+	err        error
+	rec        []byte
+	hdr        [recordHeaderSize]byte
+	buf        [pageSize]byte
+	readIndex  int   // index in buf to start at for next read
+	writeIndex int   // index in buf to start at for next write
+	total      int64 // total bytes processed.
+	index      int   // used to track partial records
+}
+
+func (r *LiveReader) Error() error {
+	return r.err
+}
+
+func (r *LiveReader) TotalRead() int64 {
+	return r.total
+}
+
+func (r *LiveReader) fillBuffer() {
+	n, err := r.rdr.Read(r.buf[r.writeIndex:len(r.buf)])
+	// We expect to get EOF, since we're reading the segment file as it's being written.
+	if err != nil && err != io.EOF {
+		r.err = err
+	}
+	r.writeIndex += n
+}
+
+// Shift the buffer up to the read index.
+func (r *LiveReader) shiftBuffer() {
+	copied := copy(r.buf[0:], r.buf[r.readIndex:r.writeIndex])
+	r.readIndex = 0
+	r.writeIndex = copied
+}
+
+func (r *LiveReader) ReadIndex() int {
+	return r.readIndex
+}
+
+func (r *LiveReader) Header() [7]byte {
+	return r.hdr
+}
+
+// Returns true if r.rec will contain a full record.
+func (r *LiveReader) Next() bool {
+	r.err = nil
+	for {
+		if r.readIndex != 0 {
+			r.shiftBuffer()
+		}
+		r.fillBuffer()
+		if r.err != nil {
+			return false
+		}
+
+		// Do we have data to read?
+		if r.writeIndex <= r.readIndex {
+			return false
+		}
+		temp, n, err := readRecord(r.buf[r.readIndex:r.writeIndex], r.hdr[:], r.total)
+		r.readIndex += n
+		r.total += int64(n)
+		if err != nil {
+			r.err = err
+			return false
+		}
+		if temp == nil {
+			return false
+		}
+
+		fullRec, err := validateRecord(recType(r.hdr[0]), r.index)
+		if err != nil {
+			r.err = err
+			r.index = 0
+			return false
+		}
+		r.rec = append(r.rec, temp...)
+		if fullRec {
+			r.index = 0
+			return true
+		}
+		// Only increment i for non-zero records since we use it
+		// to determine valid content record sequences.
+		r.index++
```

Comment:
If you pull this logic out into a separate layer, it will be much easier to follow.

Reply:
I've broken up the record reading some more, we now have `readRecord` and `buildRecord` used by `Next`.